app/src/main/java/lu/geoportail/map/LuxTileServer.kt

```diff
@@ -35,71 +47,241 @@ class LuxTileServer(
             FileOutputStream(file).use {
                 it.write(sourceFileBytes)
             }
         }
     }
 
+    private fun downloadAssetsFromMeta(json: JSONObject, toPathRoot: File) {
+        val sources = json.getJSONArray("sources")
+        for (i in 0 until sources.length()) {
+            downloadAssets(sources.getString(i), toPathRoot)
+        }
+    }
+
+    private fun downloadAssets(assetUrl: String, toPathRoot: File) {
+        val url = URL(assetUrl)
+        val filename =url.file
+        val file = File(toPathRoot, filename)
+        if(!file.parentFile.exists()) Files.createDirectories(file.parentFile.toPath())
+
+        val sourceStream = BufferedInputStream(url.openStream())
+        val fileOutputStream = FileOutputStream(file)
+        val buf = ByteArray(1024)
+        var len: Int
+        while (sourceStream.read(buf).also { len = it } > 0) {
+            fileOutputStream.write(buf, 0, len)
+        }
+        sourceStream.close()
+        fileOutputStream.close()
+    }
+
+    private fun deleteAssets(mapName: String, basePath: File) {
+        val sourcePaths = getPaths(mapName)
+        for (i in 0 until sourcePaths.length()) {
+            val url = URL(sourcePaths.getString(i))
+            val filename =url.file
+            val file = File(basePath, filename)
+            file.delete()
+        }
+    }
+
+    private fun moveAssets(sourcePaths: JSONArray, fromPath: File, toPath: File) {
+        for (i in 0 until sourcePaths.length()) {
+            val url = URL(sourcePaths.getString(i))
+            val filename =url.file
+            val toFile = File(toPath, filename)
+            if(!toFile.parentFile.exists()) Files.createDirectories(toFile.parentFile.toPath())
+            File(fromPath, filename).renameTo(toFile)
+        }
+    }
+
+    private fun getPaths(ressourceName: String): JSONArray {
+        // metadata will be read from local file until it is available online
+        return try {
+            val url = URL("file:${this.filePath}/dl/versions/$ressourceName.meta")
+            val json = JSONObject(url.openStream().use { it.readBytes() }.toString(Charsets.UTF_8))
+            val paths = json.getJSONArray("sources") ?: JSONArray()
+            paths.put("file:/versions/$ressourceName.meta")
+        } catch (e: FileNotFoundException) {
+            JSONArray()
+        }
+    }
+
+    private fun getVer(ressourceName: String): String? {
+        return try {
+            val url = URL("file:${this.filePath}/dl/versions/$ressourceName.meta")
+            val json = JSONObject(url.openStream().use { it.readBytes()}.toString(Charsets.UTF_8))
+            json.getString("version")
+        }
+        catch (e: FileNotFoundException) {
+            null
+        }
+    }
+
+    private fun getSize(ressourceName: String): String? {
+        return try {
+            val dlPath = when(getStatus(ressourceName)) {
+                "in progress" -> "tmp"
+                else -> "dl"
```

Comment:
Enum would make the pattern match avoid the "else".

Reply:
yes, you're right, I see the point now. The structure was designed the other way. First it was only used to generate a String for JSON output, then I checked the value to compute the progress.
I'll make this clearer with an enum.
Can you easily convert the enum value to String to generate the JSON output ?

Reply:
https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS42LjAiLCJwbGF0Zm9ybSI6ImphdmEiLCJhcmdzIjoiIiwibm9uZU1hcmtlcnMiOnRydWUsInRoZW1lIjoiaWRlYSIsImNvZGUiOiIvKipcbiAqIFlvdSBjYW4gZWRpdCwgcnVuLCBhbmQgc2hhcmUgdGhpcyBjb2RlLiBcbiAqIHBsYXkua290bGlubGFuZy5vcmcgXG4gKi9cbmVudW0gY2xhc3MgUkdCIHsgUkVELCBHUkVFTiwgQkxVRSB9XG5cblxuZnVuIG1haW4oKSB7XG4gICAgcHJpbnRsbihSR0IuUkVELnRvU3RyaW5nKCkpXG59In0=

Reply:
cool, thanks